### PR TITLE
[Aptos Framework][Account] added rotation capability offer and revoke functions

### PR DESCRIPTION
**Pull request by 0xchloe**
### Description
- added `offer_rotation_capability` function to allow an account owner to delegate their rotation capability to another account
- added `revoke_rotation_capability` function to allow an account owner to revoke their rotation capability offer to the `to_be_revoked_address`
- added struct `RotationCapabilityOfferProofChallengeV2` for backward compatibility and deprecated the previous `RotationCapabilityOfferProofChallenge`
- changed the more ambiguous `EACCOUNT_DOES_NOT_EXIST` error in capability-related functions to either `EOFFERER_ADDRESS_DOES_NOT_EXIST` or `ERECIPIENT_ADDRESS_DOES_NOT_EXIST`
- changed `ENO_SUCH_SIGNER_CAPABILITY` error to the more accurate naming `ENO_SUCH_SIGNER_CAPABILITY_OFFER`
- changed sequence number for `harness.new_account_with_key_pair()` from 10 to 0, so if other devs are trying to generate a valid signature for unit test purpose, they don't have to manually change the sequence number from 10 to 0. 

### Test Plan

unit & e2e tests

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/4202)

### CODE REVIEW

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 1]
```diff
@@ -292,6 +292,51 @@ module aptos_framework::account {
...
+        account_public_key_bytes: vector<u8>,
+        recipient_address: address,
+    ) acquires Account {
+        let addr = signer::address_of(account);
+        assert!(exists_at(addr) && exists_at(recipient_address), error::not_found(EACCOUNT_DOES_NOT_EXIST));
+
+        let account_resource = borrow_global_mut<Account>(addr);
+
+        // proof that this account intends to delegate its rotation capability to another account
+        let proof_challenge = RotationCapabilityOfferProofChallenge {
```
Comment (alinush) [inline aptos-move/framework/aptos-framework/sources/account.move:308]: Add another field for the address of the source account here.

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 2]
```diff
@@ -292,6 +292,51 @@ module aptos_framework::account {
...
+            sequence_number: account_resource.sequence_number,
+            recipient_address,
+        };
+
+        // verify that the `SignerCapabilityOfferProofChallenge` is correct and signed by the account owner's private key
```
Comment (alinush) [inline aptos-move/framework/aptos-framework/sources/account.move:313]: Typo: SignerCapabilityOfferProofChallenge should be RotationCapabilityOfferProofChallenge

Comment (alinush) [reply aptos-move/framework/aptos-framework/sources/account.move:313]: Technically, you don't "verify correctness of RotationCapabilityOfferProofChallenge" here. You've just built the correct one above, so I would rephrase this as: "verify the signature on RotationCapabilityOfferProofChallenge by the account owner"

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 3]
```diff
@@ -292,6 +292,51 @@ module aptos_framework::account {
...
+            assert!(account_resource.authentication_key == expected_auth_key, error::invalid_argument(EWRONG_CURRENT_PUBLIC_KEY));
+
+            let rotation_capability_sig = multi_ed25519::new_signature_from_bytes(rotation_capability_sig_bytes);
+            assert!(multi_ed25519::signature_verify_strict_t(&rotation_capability_sig, &pubkey, proof_challenge), error::invalid_argument(EINVALID_PROOF_OF_KNOWLEDGE));
+        } else {
+            abort error::invalid_argument(EINVALID_SCHEME)
+        };
+
+        // update the existing rotation capability offer or put in a new rotation capability offer for the current account
+        if (option::is_some(&account_resource.rotation_capability_offer.for)) {
```
Comment (alinush) [inline aptos-move/framework/aptos-framework/sources/account.move:333]: Why not use `std::option::swap_or_fill` here?

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 4]
```diff
@@ -531,8 +576,17 @@ module aptos_framework::account {
...
-/*
-TODO bring back with generic rotation capability
+    #[test(bob = @0x345)]
+    #[expected_failure(abort_code = 65544)]
+    public entry fun test_valid_offer_rotation_capability(bob: signer) acquires Account {
+        let pk = x"f66bf0ce5ceb582b93d6780820c2025b9967aedaa259bdbb9f3d0297eced0e18";
+        let alice = create_account_from_ed25519_public_key(pk);
+        create_account(signer::address_of(&bob));
+
+        let valid_signature = x"cd181d65eb31193dcf1627fc0cc04208f66e7f243facc840830eaa458b176de570f73b661c127d98bc276c5a07ab242734b4d656163a86803561c0b9d9d01d0c";
```
Comment (alinush) [inline aptos-move/framework/aptos-framework/sources/account.move:586]: Same issue as in the other PR: might be good to add instructions on how you generated this signature.

File: aptos-move/framework/aptos-framework/sources/account.move (hunk 1 above)
Comment (chloeqjz) [reply aptos-move/framework/aptos-framework/sources/account.move:308]: cc @davidiw is this a breaking change?

File: aptos-move/framework/aptos-framework/sources/account.move (hunk 2 above)
Comment (chloeqjz) [reply aptos-move/framework/aptos-framework/sources/account.move:313]: ok will update the comment! good point

File: aptos-move/framework/aptos-framework/sources/account.move (hunk 3 above)
Comment (chloeqjz) [reply aptos-move/framework/aptos-framework/sources/account.move:333]: ahhhh didn't know that exists! I was only looking at [this move guide](https://move-language.github.io/move/standard-library.html?highlight=option#option). is there an Aptos-specific move guide?

Comment (alinush) [reply aptos-move/framework/aptos-framework/sources/account.move:333]: Hehe, nope! I was able to find it in `aptos-core/aptos-move/framework/move-stdlib/sources/option.move`.

Be sure to look there for Move standard library features.

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 5]
```diff
@@ -292,6 +292,51 @@ module aptos_framework::account {
...
 
+    public entry fun offer_rotation_capability(
+        account: &signer,
+        rotation_capability_sig_bytes: vector<u8>,
+        account_scheme: u8,
+        account_public_key_bytes: vector<u8>,
+        recipient_address: address,
+    ) acquires Account {
+        let addr = signer::address_of(account);
+        assert!(exists_at(addr) && exists_at(recipient_address), error::not_found(EACCOUNT_DOES_NOT_EXIST));
```
Comment (areshand) [inline aptos-move/framework/aptos-framework/sources/account.move:303]: exists_at(addr) not necessary?

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 6]
```diff
@@ -531,8 +576,17 @@ module aptos_framework::account {
...
-TODO bring back with generic rotation capability
+    #[test(bob = @0x345)]
+    #[expected_failure(abort_code = 65544)]
+    public entry fun test_valid_offer_rotation_capability(bob: signer) acquires Account {
+        let pk = x"f66bf0ce5ceb582b93d6780820c2025b9967aedaa259bdbb9f3d0297eced0e18";
+        let alice = create_account_from_ed25519_public_key(pk);
+        create_account(signer::address_of(&bob));
+
+        let valid_signature = x"cd181d65eb31193dcf1627fc0cc04208f66e7f243facc840830eaa458b176de570f73b661c127d98bc276c5a07ab242734b4d656163a86803561c0b9d9d01d0c";
+        offer_rotation_capability(&alice, valid_signature, 0, pk, signer::address_of(&bob));
```
Comment (areshand) [inline aptos-move/framework/aptos-framework/sources/account.move:587]: could we add assert to show bob has the capability?

File: aptos-move/framework/aptos-framework/sources/account.move (hunk 5 above)
Comment (alinush) [reply aptos-move/framework/aptos-framework/sources/account.move:303]: It might be possible there isn't an `Account` resource there though...

Comment (davidiw) [reply aptos-move/framework/aptos-framework/sources/account.move:303]: it isn't possible in our current framework to create a signer without an account.

but it isn't great to have the same error for two different errors

File: aptos-move/framework/aptos-framework/sources/account.move (hunk 1 above)
Comment (davidiw) [reply aptos-move/framework/aptos-framework/sources/account.move:308]: well it isn't being used, so we could try to amend it and see if you can pass the `e2e-move-tests` for upgradeability
it isn't strictly necessary since the public key already ties the proof to the account but it is a good practice

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 7]
```diff
@@ -292,6 +292,51 @@ module aptos_framework::account {
...
+        } else if (account_scheme == MULTI_ED25519_SCHEME) {
+            let pubkey = multi_ed25519::new_unvalidated_public_key_from_bytes(account_public_key_bytes);
+            let expected_auth_key = multi_ed25519::unvalidated_public_key_to_authentication_key(&pubkey);
+            assert!(account_resource.authentication_key == expected_auth_key, error::invalid_argument(EWRONG_CURRENT_PUBLIC_KEY));
+
+            let rotation_capability_sig = multi_ed25519::new_signature_from_bytes(rotation_capability_sig_bytes);
+            assert!(multi_ed25519::signature_verify_strict_t(&rotation_capability_sig, &pubkey, proof_challenge), error::invalid_argument(EINVALID_PROOF_OF_KNOWLEDGE));
+        } else {
+            abort error::invalid_argument(EINVALID_SCHEME)
+        };
```
Comment (davidiw) [inline aptos-move/framework/aptos-framework/sources/account.move:330]: this feels like a lot of reusable boilerplate code... @alinush are you going to make a general library or should we just add that?

File: aptos-move/framework/aptos-framework/sources/account.move (hunk 3 above)
Comment (davidiw) [reply aptos-move/framework/aptos-framework/sources/account.move:333]: yes, there is a great aptos-specific move guide... the code :trollface:

File: aptos-move/framework/aptos-framework/sources/account.move (hunk 4 above)
Comment (davidiw) [reply aptos-move/framework/aptos-framework/sources/account.move:586]: Personally I don't think we should or need to validate that our crypto libraries are doing the right thing. I do agree that if we break something (which we aren't allowed to any more), then having gen libraries is great

**Review by davidiw — COMMENTED** [review]
I'm good, but you have two active reviewers, so I'll let them approve but poke me if need be :)

File: aptos-move/framework/aptos-framework/sources/account.move (hunk 7 above)
Comment (alinush) [reply aptos-move/framework/aptos-framework/sources/account.move:330]: Yes, I think we could abstract this either using this `account_scheme` parameter or with generics. Perhaps in a new `aptos_std::signatures` module?

File: aptos-move/framework/aptos-framework/sources/account.move (hunk 4 above)
Comment (alinush) [reply aptos-move/framework/aptos-framework/sources/account.move:586]: What do you mean by "gen libraries"?

**Review by davidiw — CHANGES_REQUESTED** [review]
PR hasn't been updated for a bit... we should get this out before first update

File: aptos-move/framework/aptos-framework/sources/account.move (hunk 1 above)
Comment (chloeqjz) [reply aptos-move/framework/aptos-framework/sources/account.move:334]: just added the address of the source account - do we want to do the same for the `SignerCapabilityOfferProofChallenge` (which would be a breaking change)?

File: aptos-move/framework/aptos-framework/sources/account.move (hunk 7 above)
Comment (chloeqjz) [reply aptos-move/framework/aptos-framework/sources/account.move:357]: @alinush are you going to make a gen lib or should we do that? if we do that, would you happen to have an example?

Comment (alinush) [reply aptos-move/framework/aptos-framework/sources/account.move:373]: That will be a broader effort that should include domain separation too. It'll take more time...

**Comment by areshand** [comment]
I don't see any function using rotation capability to rotate the auth key. we should add these function after the capability is used somewhere

**Review by davidiw — COMMENTED** [review]
@chloeqjz / @alinush is the current plan to wait until we have the new domain separation?

**Comment by chloeqjz** [comment]
> @chloeqjz / @alinush is the current plan to wait until we have the new domain separation?

yes! we communicated on slack

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 8]
```diff
@@ -64,6 +64,15 @@ module aptos_framework::account {
         new_public_key: vector<u8>,
     }
 
+    /// This struct stores the challenge message that should be signed by the source account, when the source account
+    /// is delegating its rotation capability to the `recipient_address`.
+    struct RotationCapabilityOfferProofChallengeV2 has drop {
+        sequence_number: u64,
+        source_address: address,
```
Comment (lightmark) [inline aptos-move/framework/aptos-framework/sources/account.move:71]: Could you comment why `source_address` is necessary here?

File: aptos-move/e2e-move-tests/src/harness.rs [hunk 9]
```diff
@@ -109,9 +109,9 @@ impl MoveHarness {
         let privkey = Ed25519PrivateKey::generate(&mut rng);
         let pubkey = privkey.public_key();
         let acc = Account::with_keypair(privkey, pubkey);
-        let data = AccountData::with_account(acc.clone(), 1_000_000_000_000_000, 10);
+        let data = AccountData::with_account(acc.clone(), 1_000_000_000_000_000, 0);
         self.executor.add_account_data(&data);
-        self.txn_seq_no.insert(*acc.address(), 10);
```
Comment (lightmark) [inline aptos-move/e2e-move-tests/src/harness.rs:114]: Why does this change relate to the PR.

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 10]
```diff
@@ -311,15 +326,74 @@ module aptos_framework::account {
...
+        } else {
+            abort error::invalid_argument(EINVALID_SCHEME)
+        };
+
+        // update the existing rotation capability offer or put in a new rotation capability offer for the current account
+        option::swap_or_fill(&mut account_resource.rotation_capability_offer.for, recipient_address);
+    }
+
+    /// Revoke the rotation capability offer given to `to_be_revoked_recipient_address` from `account`
+    public entry fun revoke_rotation_capability(account: &signer, to_be_revoked_address: address) acquires Account {
```
Comment (lightmark) [inline aptos-move/framework/aptos-framework/sources/account.move:380]: why is `to_be_revoked_address` required? It seems simple w/o that.

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 11]
```diff
@@ -328,7 +402,7 @@ module aptos_framework::account {
         recipient_address: address
     ) acquires Account {
         let source_address = signer::address_of(account);
-        assert!(exists_at(source_address) && exists_at(recipient_address), error::not_found(EACCOUNT_DOES_NOT_EXIST));
+        assert!(exists_at(recipient_address), error::not_found(ERECIPIENT_ADDRESS_DOES_NOT_EXIST));
```
Comment (lightmark) [inline aptos-move/framework/aptos-framework/sources/account.move:405]: why delete `exists_at(source_address)`
